File: 1143.最长公共子序列/solution.go
func longestCommonSubsequence(text1 string, text2 string) int {
	sz1, sz2 := len(text1), len(text2)
	dp := make([][]int, sz1)
	for i := range dp {
		dp[i] = make([]int, sz2)
	}
	if text1[0] == text2[0] {
		dp[0][0] = 1
	}
	for j := 1; j < sz2; j++ {
		if text1[0] == text2[j] {
			dp[0][j] = 1
		} else {
			dp[0][j] = dp[0][j-1]
		}
	}
	for i := 1; i < sz1; i++ {
		if text1[i] == text2[0] {
			dp[i][0] = 1
		} else {
			dp[i][0] = dp[i-1][0]
		}
	}
	for i := 1; i < sz1; i++ {
		for j := 1; j < sz2; j++ {
			if text1[i] == text2[j] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
		}
	}
	return dp[sz1-1][sz2-1]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 1143.最长公共子序列/solution.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int sz1=text1.size(),sz2=text2.size();
        vector<vector<int>> dp(sz1,vector<int>(sz2,0));
        dp.at(0).at(0)= text1.at(0)==text2.at(0) ? 1 : 0;
        for (int j=1;j<sz2;j++) {
            if (text1.at(0)==text2.at(j)) {
                dp.at(0).at(j)=1;
            } else {
                dp.at(0).at(j)=dp.at(0).at(j-1);
            }
        }
        for (int i=1;i<sz1;i++) {
            if (text1.at(i)==text2.at(0)) {
                dp.at(i).at(0)=1;
            } else {
                dp.at(i).at(0)=dp.at(i-1).at(0);
            }
        }
        for (int i=1;i<sz1;i++) {
            for (int j=1;j<sz2;j++) {
                if (text1.at(i)==text2.at(j)) {
                    dp.at(i).at(j)=max(dp.at(i-1).at(j-1)+1,max(dp.at(i-1).at(j),dp.at(i).at(j-1)));
                } else {
                    dp.at(i).at(j)=max(dp.at(i-1).at(j),dp.at(i).at(j-1));
                }
            }
        }
        return dp.at(sz1-1).at(sz2-1);
    }
};